### PR TITLE
filter bar full functionality

### PR DESCRIPTION
**Pull request by cjmaret**
# Describe your changes 

At this PR I ... 

1. on clicking Apply, the filter params are sent to the backend, filtering the startup cards, and then getting the array of matching cards (and setting localStorage)
2. on clicking Clear, the displayed variables are cleared, an empty API call is sent, all startups are retrieved, and localStorage is set
3. on clicking Other, displayed variables are cleared, empty API call is sent, and localStorage is NOT set
4. on clicking Relevant, data from localStorage is retrieved, and API call with the stored parameters is sent
5. minor display bug fixes

Loom with explanation: 
https://www.loom.com/share/b4ef6700cf8d41618a1dec0a59971ac8

[startup-card-data.json.zip](https://github.com/vgalilei/venturecat/files/7633242/startup-card-data.json.zip)

EDIT: I left a JSON file above, you can grab that, and import it into your Compass collection. It will help to have the collection be empty, because sometimes you get crazy "duplicate key" errors if there's data in the collection.

 Once you copy them, it will add that ObjectId at the top, and you can manually change the values for revenue, segment, location, and growth if you want to see the different cards appear when you change the filter inputs.
Screen Shot 2021-11-29 at 1 14 59 AM

Also, I made this change to the server, to reach the database. I'm not sure if this can stay like this, or it needs to change, but the original mongoose.connect is the one commented out, and the one i wrote is the second one.
Screen Shot 2021-11-29 at 1 28 39 AM

Here is what the collections look like in Compass, a venturecat folder with investors and startups subfolder, that's what the server can connect to.
Screen Shot 2021-11-29 at 1 29 58 AM

# I follow best practices 

- [ ] My code is covered by unit tests 
- [ ] This PR is less than [300 lines of code](https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests)
- [ ] My code [smells good](https://en.wikipedia.org/wiki/Code_smell) 
- [ ] I have tested my PR locally and proven it works. 
- [ ] There is [no copy-paste](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

### CODE REVIEW

**Comment by McCambley** [comment]
@cjmaret 

> Also, I made this change to the server, to reach the database. I'm not sure if this can stay like this, or it needs to change, but the original mongoose.connect is the one commented out, and the one i wrote is the second one.
Keep it as is, `mongoose.connect(MONGO_URI)`, as that connects with the `.env` variable

File: frontend/src/components/investor-feed/InvestorFeed.js [hunk 1]
```diff
@@ -22,70 +23,171 @@ function InvestorFeed() {
     min: 10,
     max: 200,
   });
-  const [storedValues, setStoredValues] = React.useState({
+  const [storedStartupFilterValues, setStoredStartupFilterValues] = React.useState({
```
Comment (McCambley) [inline frontend/src/components/investor-feed/InvestorFeed.js:26]: Name change is 👍 thank you

File: frontend/src/components/investor-feed/InvestorFeed.js [hunk 2]
```diff
@@ -22,70 +23,171 @@ function InvestorFeed() {
...
+  React.useEffect(() => {
+    const revenue = storedStartupFilterValues.revenueRangeInputState;
+    const segment = storedStartupFilterValues.selectInputState.segment;
+    const location = storedStartupFilterValues.selectInputState.location;
+    const growth = storedStartupFilterValues.growthRangeInputState;
+
+    setStartupCardData(null);
+    setIsLoadingCards(true);
+    api
+      .getStartupCards(revenue, segment, location, growth, user /*token*/)
```
Comment (McCambley) [inline frontend/src/components/investor-feed/InvestorFeed.js:66]: Token shouldn't be needed.

File: frontend/src/components/investor-feed/InvestorFeed.js [hunk 3]
```diff
@@ -22,70 +23,171 @@ function InvestorFeed() {
...
+      .then((res) => {
+        setTimeout(() => {
+          setIsLoadingCards(false);
+          setStartupCardData(res);
+        }, 1000);
+      })
+      .catch((err) => console.log(err));
```
Comment (McCambley) [inline frontend/src/components/investor-feed/InvestorFeed.js:73]: console.error(err)

File: frontend/src/components/investor-feed/InvestorFeed.js [hunk 4]
```diff
@@ -22,70 +23,171 @@ function InvestorFeed() {
...
-  }
+    let revenue;
+    let segment;
+    let location;
+    let growth;
 
-  // get the full array of startup cards
-  React.useEffect(() => {
+    // send API with default data, or with data set by user
+    if (clearButtonRefClicked.current || otherButtonRefClicked.current) {
```
Comment (McCambley) [inline frontend/src/components/investor-feed/InvestorFeed.js:100]: Connecting this comment to the idea of doing the one initial request of startups and filtering on the frontend. This conditional speedbump wouldn't be needed in that case, as there'd be no subsequent Api call when the "Clear" or "Other" button is called, we'd just be displaying more of the array of startups that has already been fetched in the one previous api call.

File: frontend/src/components/investor-feed/InvestorFeed.js [hunk 5]
```diff
@@ -22,70 +23,171 @@ function InvestorFeed() {
...
+  // get the initial array of startup cards
+  React.useEffect(() => {
+    const revenue = storedStartupFilterValues.revenueRangeInputState;
+    const segment = storedStartupFilterValues.selectInputState.segment;
+    const location = storedStartupFilterValues.selectInputState.location;
+    const growth = storedStartupFilterValues.growthRangeInputState;
+
+    setStartupCardData(null);
+    setIsLoadingCards(true);
+    api
```
Comment (McCambley) [inline frontend/src/components/investor-feed/InvestorFeed.js:65]: This is repeated on `120` and should be stripped out to a singular function if possible

**Review by McCambley — COMMENTED** [review]
Hey Colin, 
Awesome work here. It's all working as described, well done.

I've left a few comments below, and a few of them need adressing. 

Some notes: 
1) If you want, you can export your MongoDB data from compass and attach a JSON file to your PR. This might make testing easier for those that don't have your data. 
2) I didn't comment on what we had talked about in the PM about using promises to make things work. I can't seem to find where that happens or what the problem was and how you solved it. would you be able to elaborate on that on slack some how? I do think that it'd be easier to make one initial massive request of startups and then filter them on the frontend. If we do that, I think that would prevent the need for promises, but that would require a lot of refactoring so I want to chat about that with you and the group at a meeting.
3) There's a comment about the MONGO_URI that needs to stay, mentioning it here just to highlight it.

Well done with all of this, it seemed like no easy task to get it all to work.

File: backend/routes/index.js [hunk 6]
```diff
@@ -11,7 +11,11 @@ const startupsRouter = require('./startups');
 rootRouter.use('/api/auth', authRoute);
 
 // Routes protected with middleware
-rootRouter.use('/api/investors', ensureAuthenticated, investorRouter);
-rootRouter.use('/api/startups', ensureAuthenticated, startupsRouter);
+rootRouter.use('/api/investors',
+// ensureAuthenticated,
```
[the remark this answers is not in the document]
Comment (cjmaret) [reply backend/routes/index.js:15]: ah sorry, it was (and still is) causing a CORS error, let me fix that though

File: backend/routes/index.js [hunk 7]
```diff
@@ -1,5 +1,5 @@
 const express = require('express');
-const { ensureAuthenticated } = require('../middleware/auth');
+// const { ensureAuthenticated } = require('../middleware/auth');
```
Comment (vgalilei) [inline backend/routes/index.js:2]: Remember to comment it out before merge

**Comment by vgalilei** [comment]
Overall LGTM 
* Please revert back authentification (better just use .env variables to test locally) 
* For segments and regions it is better to introduce identifiers and then convert them to the String representation on Frontend. I would like to decouple UI strings that can be changed from database objects that can't be changed easily

File: frontend/src/components/investor-feed/InvestorFeed.js [hunk 8]
```diff
@@ -22,72 +23,193 @@ function InvestorFeed() {
...
+      min: 10,
+      max: 200,
+    });
+    otherButtonRefClicked.current = true;
+    getStartupCards();
+    otherButtonRefClicked.current = false;
+  }
+
+  // only shows Profile component if filter bar hasn't been used AND there are no cards
+  function usingFilter() {
```
Comment (McCambley) [inline frontend/src/components/investor-feed/InvestorFeed.js:201]: Nice!

**Comment by McCambley** [comment]
Love the usingFilter thing, I'm don't fully understand it though. Looks like when you either have no startups or when you click clear the investor feed will hide and you'll see the profile view?

**Review by McCambley — COMMENTED** [review]
Love the usingFilter thing, I'm don't fully understand it though. Looks like when you either have no startups or when you click clear the investor feed will hide and you'll see the profile view? Other than my lack of understanding here, looks good to me!

**Review by McCambley — APPROVED** [review]
I understand now 👍

**Pull request merged**